File: skills/references/realtime-streams.py
"""
Pattern: Realtime Streams
Comparable to: Socket.io, Pusher, Firebase Realtime

Push live data to connected WebSocket clients.
Clients connect at: ws://host:3112/stream/{stream_name}/{group_id}

Built-in stream operations: stream::set, stream::get, stream::list,
stream::delete, stream::send.

Note: The Python SDK does not support createStream for custom adapters.
Use the built-in stream operations and state-backed presence instead.

How-to references:
  - Realtime streams: https://iii.dev/docs/how-to/stream-realtime-data
"""

import asyncio
import os
import uuid
from datetime import datetime, timezone

from iii import InitOptions, Logger, TriggerAction, register_worker

iii = register_worker(
    address=os.environ.get("III_ENGINE_URL", "ws://localhost:49134"),
    options=InitOptions(worker_name="realtime-streams"),
)

# ---
# stream::set — Persist an item in a stream group
# Payload: { stream_name, group_id, item_id, data }
# ---


async def post_message(data):
    logger = Logger()
    message_id = f"msg-{uuid.uuid4().hex}"

    await iii.trigger_async({
        "function_id": "stream::set",
        "payload": {
            "stream_name": "chat",
            "group_id": data["room"],
            "item_id": message_id,
            "data": {
                "sender": data["sender"],
                "text": data["text"],
                "timestamp": datetime.now(timezone.utc).isoformat(),
            },
        },
    })

    logger.info("Message stored in stream", {"room": data["room"], "messageId": message_id})
    return {"messageId": message_id}


iii.register_function("chat::post-message", post_message)

# ---
# stream::get — Retrieve a single item from a stream group
# Payload: { stream_name, group_id, item_id }
# ---


async def get_message(data):
    message = await iii.trigger_async({
        "function_id": "stream::get",
        "payload": {
            "stream_name": "chat",
            "group_id": data["room"],
            "item_id": data["messageId"],
        },
    })

    if not message:
        return {"error": "Message not found"}

    return message


iii.register_function("chat::get-message", get_message)

# ---
# stream::list — List all items in a stream group
# Payload: { stream_name, group_id }
# ---


async def list_messages(data):
    messages = await iii.trigger_async({
        "function_id": "stream::list",
        "payload": {
            "stream_name": "chat",
            "group_id": data["room"],
        },
    })

    return {"room": data["room"], "messages": messages or []}


iii.register_function("chat::list-messages", list_messages)

# ---
# stream::delete — Remove an item from a stream group
# Payload: { stream_name, group_id, item_id }
# ---


async def delete_message(data):
    await iii.trigger_async({
        "function_id": "stream::delete",
        "payload": {
            "stream_name": "chat",
            "group_id": data["room"],
            "item_id": data["messageId"],
        },
    })

    return {"deleted": data["messageId"]}


iii.register_function("chat::delete-message", delete_message)

# ---
# stream::send — Push a live event to all connected clients
# Clients on ws://host:3112/stream/chat/{room} receive this instantly.
# Use TriggerAction.Void() for fire-and-forget delivery.
# ---


async def broadcast(data):
    logger = Logger()
    event_id = f"evt-{uuid.uuid4().hex}"

    await iii.trigger_async({
        "function_id": "stream::set",
        "payload": {
            "stream_name": "chat",
            "group_id": data["room"],
            "item_id": event_id,
            "data": {
                "sender": data["sender"],
                "text": data["text"],
                "timestamp": datetime.now(timezone.utc).isoformat(),
            },
        },
    })

    iii.trigger({
        "function_id": "stream::send",
        "payload": {
            "stream_name": "chat",
            "group_id": data["room"],
            "id": event_id,
            "event_type": "new_message",
            "data": {
                "sender": data["sender"],
                "text": data["text"],
                "timestamp": datetime.now(timezone.utc).isoformat(),
            },
        },
        "action": TriggerAction.Void(),
    })

    logger.info("Message broadcast", {"room": data["room"], "eventId": event_id})
    return {"eventId": event_id}


iii.register_function("chat::broadcast", broadcast)

# ---
# Presence tracking — user joins/leaves
# Uses state-backed storage since Python SDK lacks createStream.
# Clients connect at: ws://host:3112/stream/presence/{room}
# ---


async def presence_join(data):
    await iii.trigger_async({
        "function_id": "state::set",
        "payload": {
            "scope": f"presence::{data['room']}",
            "key": data["userId"],
            "value": {
                "userId": data["userId"],
                "name": data["name"],
                "status": "online",
                "updated_at": datetime.now(timezone.utc).isoformat(),
            },
        },
    })

    iii.trigger({
        "function_id": "stream::send",
        "payload": {
            "stream_name": "presence",
            "group_id": data["room"],
            "id": f"join-{uuid.uuid4().hex}",
            "event_type": "user_joined",
            "data": {"userId": data["userId"], "name": data["name"]},
        },
        "action": TriggerAction.Void(),
    })

    return {"joined": data["room"]}


iii.register_function("presence::join", presence_join)


async def presence_leave(data):
    await iii.trigger_async({
        "function_id": "state::delete",
        "payload": {
            "scope": f"presence::{data['room']}",
            "key": data["userId"],
        },
    })

    iii.trigger({
        "function_id": "stream::send",
        "payload": {
            "stream_name": "presence",
            "group_id": data["room"],
            "id": f"leave-{uuid.uuid4().hex}",
            "event_type": "user_left",
            "data": {"userId": data["userId"]},
        },
        "action": TriggerAction.Void(),
    })

    return {"left": data["room"]}


iii.register_function("presence::leave", presence_leave)

# ---
# HTTP triggers
# ---
iii.register_trigger({"type": "http", "function_id": "chat::broadcast", "config": {"api_path": "/chat/send", "http_method": "POST"}})
iii.register_trigger({"type": "http", "function_id": "chat::list-messages", "config": {"api_path": "/chat/:room/messages", "http_method": "GET"}})
iii.register_trigger({"type": "http", "function_id": "presence::join", "config": {"api_path": "/presence/join", "http_method": "POST"}})
iii.register_trigger({"type": "http", "function_id": "presence::leave", "config": {"api_path": "/presence/leave", "http_method": "POST"}})


async def main():
    while True:
        await asyncio.sleep(60)


if __name__ == "__main__":
    asyncio.run(main())

File: skills/references/realtime-streams.js
/**
 * Pattern: Realtime Streams
 * Comparable to: Socket.io, Pusher, Firebase Realtime
 *
 * Push live data to connected WebSocket clients.
 * Clients connect at: ws://host:3112/stream/{stream_name}/{group_id}
 *
 * Built-in stream operations: stream::set, stream::get, stream::list,
 * stream::delete, stream::send. Use createStream for custom adapters.
 *
 * How-to references:
 *   - Realtime streams: https://iii.dev/docs/how-to/stream-realtime-data
 */

import { registerWorker, Logger, TriggerAction } from 'iii-sdk'

const iii = registerWorker(process.env.III_ENGINE_URL || 'ws://localhost:49134', {
  workerName: 'realtime-streams',
})

// ---------------------------------------------------------------------------
// stream::set — Persist an item in a stream group
// Payload: { stream_name, group_id, item_id, data }
// ---------------------------------------------------------------------------
iii.registerFunction('chat::post-message', async (data) => {
  const logger = new Logger()
  const messageId = `msg-${Date.now()}`

  await iii.trigger({
    function_id: 'stream::set',
    payload: {
      stream_name: 'chat',
      group_id: data.room,
      item_id: messageId,
      data: {
        sender: data.sender,
        text: data.text,
        timestamp: new Date().toISOString(),
      },
    },
  })

  logger.info('Message stored in stream', { room: data.room, messageId })
  return { messageId }
})

// ---------------------------------------------------------------------------
// stream::get — Retrieve a single item from a stream group
// Payload: { stream_name, group_id, item_id }
// ---------------------------------------------------------------------------
iii.registerFunction('chat::get-message', async (data) => {
  const message = await iii.trigger({
    function_id: 'stream::get',
    payload: {
      stream_name: 'chat',
      group_id: data.room,
      item_id: data.messageId,
    },
  })

  if (!message) {
    return { error: 'Message not found' }
  }

  return message
})

// ---------------------------------------------------------------------------
// stream::list — List all items in a stream group
// Payload: { stream_name, group_id }
// ---------------------------------------------------------------------------
iii.registerFunction('chat::list-messages', async (data) => {
  const messages = await iii.trigger({
    function_id: 'stream::list',
    payload: {
      stream_name: 'chat',
      group_id: data.room,
    },
  })

  return { room: data.room, messages: messages || [] }
})

// ---------------------------------------------------------------------------
// stream::delete — Remove an item from a stream group
// Payload: { stream_name, group_id, item_id }
// ---------------------------------------------------------------------------
iii.registerFunction('chat::delete-message', async (data) => {
  await iii.trigger({
    function_id: 'stream::delete',
    payload: {
      stream_name: 'chat',
      group_id: data.room,
      item_id: data.messageId,
    },
  })

  return { deleted: data.messageId }
})

// ---------------------------------------------------------------------------
// stream::send — Push a live event to all connected clients
// Clients on ws://host:3112/stream/chat/{room} receive this instantly.
// Use TriggerAction.Void() for fire-and-forget delivery.
// ---------------------------------------------------------------------------
iii.registerFunction('chat::broadcast', async (data) => {
  const logger = new Logger()
  const eventId = `evt-${Date.now()}`

  // Store the message
  await iii.trigger({
    function_id: 'stream::set',
    payload: {
      stream_name: 'chat',
      group_id: data.room,
      item_id: eventId,
      data: {
        sender: data.sender,
        text: data.text,
        timestamp: new Date().toISOString(),
      },
    },
  })

  // Push live event to connected WebSocket clients (fire-and-forget)
  iii.trigger({
    function_id: 'stream::send',
    payload: {
      stream_name: 'chat',
      group_id: data.room,
      id: eventId,
      event_type: 'new_message',
      data: {
        sender: data.sender,
        text: data.text,
        timestamp: new Date().toISOString(),
      },
    },
    action: TriggerAction.Void(),
  })

  logger.info('Message broadcast', { room: data.room, eventId })
  return { eventId }
})

// ---------------------------------------------------------------------------
// createStream — Custom stream adapter with get/set/delete/list/listGroups
// Useful for integrating external data sources as stream backends.
// ---------------------------------------------------------------------------
iii.createStream('presence', {
  get: async ({ group_id, item_id }) => {
    return await iii.trigger({
      function_id: 'state::get',
      payload: { scope: `presence::${group_id}`, key: item_id },
    })
  },
  set: async ({ group_id, item_id, data }) => {
    await iii.trigger({
      function_id: 'state::set',
      payload: {
        scope: `presence::${group_id}`,
        key: item_id,
        value: { ...data, updated_at: new Date().toISOString() },
      },
    })

    // Maintain presence registry so listGroups() returns accurate data
    const registry = await iii.trigger({ function_id: 'state::get', payload: { scope: 'presence-registry', key: 'groups' } })
    const groups = registry?.groups || []
    if (!groups.includes(group_id)) {
      groups.push(group_id)
      await iii.trigger({ function_id: 'state::set', payload: { scope: 'presence-registry', key: 'groups', value: { groups } } })
    }
  },
  delete: async ({ group_id, item_id }) => {
    await iii.trigger({
      function_id: 'state::delete',
      payload: { scope: `presence::${group_id}`, key: item_id },
    })

    // Remove empty groups from registry
    const remaining = await iii.trigger({ function_id: 'state::list', payload: { scope: `presence::${group_id}` } })
    if (!remaining || remaining.length === 0) {
      const registry = await iii.trigger({ function_id: 'state::get', payload: { scope: 'presence-registry', key: 'groups' } })
      const groups = (registry?.groups || []).filter(g => g !== group_id)
      await iii.trigger({ function_id: 'state::set', payload: { scope: 'presence-registry', key: 'groups', value: { groups } } })
    }
  },
  list: async ({ group_id }) => {
    return await iii.trigger({
      function_id: 'state::list',
      payload: { scope: `presence::${group_id}` },
    })
  },
  listGroups: async () => {
    const registry = await iii.trigger({
      function_id: 'state::get',
      payload: { scope: 'presence-registry', key: 'groups' },
    })
    return registry?.groups || []
  },
})

// ---------------------------------------------------------------------------
// Presence tracking — user joins/leaves
// Clients connect at: ws://host:3112/stream/presence/{room}
// ---------------------------------------------------------------------------
iii.registerFunction('presence::join', async (data) => {
  await iii.trigger({
    function_id: 'stream::set',
    payload: {
      stream_name: 'presence',
      group_id: data.room,
      item_id: data.userId,
      data: { userId: data.userId, name: data.name, status: 'online' },
    },
  })

  // Notify all connected clients
  iii.trigger({
    function_id: 'stream::send',
    payload: {
      stream_name: 'presence',
      group_id: data.room,
      id: `join-${Date.now()}`,
      event_type: 'user_joined',
      data: { userId: data.userId, name: data.name },
    },
    action: TriggerAction.Void(),
  })

  return { joined: data.room }
})

iii.registerFunction('presence::leave', async (data) => {
  await iii.trigger({
    function_id: 'stream::delete',
    payload: {
      stream_name: 'presence',
      group_id: data.room,
      item_id: data.userId,
    },
  })

  iii.trigger({
    function_id: 'stream::send',
    payload: {
      stream_name: 'presence',
      group_id: data.room,
      id: `leave-${Date.now()}`,
      event_type: 'user_left',
      data: { userId: data.userId },
    },
    action: TriggerAction.Void(),
  })

  return { left: data.room }
})

// ---------------------------------------------------------------------------
// HTTP triggers
// ---------------------------------------------------------------------------
iii.registerTrigger({ type: 'http', function_id: 'chat::broadcast', config: { api_path: '/chat/send', http_method: 'POST' } })
iii.registerTrigger({ type: 'http', function_id: 'chat::list-messages', config: { api_path: '/chat/:room/messages', http_method: 'GET' } })
iii.registerTrigger({ type: 'http', function_id: 'presence::join', config: { api_path: '/presence/join', http_method: 'POST' } })
iii.registerTrigger({ type: 'http', function_id: 'presence::leave', config: { api_path: '/presence/leave', http_method: 'POST' } })
